pkg/kubectl/cmd/auth/cani.go

```diff
@@ -196,6 +200,8 @@ func (o *CanIOptions) Validate() error {
 		if o.Resource != (schema.GroupVersionResource{}) || o.ResourceName != "" {
 			return fmt.Errorf("NonResourceURL and ResourceName can not specified together")
 		}
+	} else {
+		o.checkResource()
```

Comment:
put more of the conditions that would keep us from running this check here, and make the function to determine whether the particular resource is namespaced more constrained in its input, like this:
```
} else if !o.Resource.Empty() && !o.AllNamespaces && o.DiscoveryClient != nil {
  if namespaced, err := isNamespaced(o.Resource, o.DiscoveryClient); err == nil && !namespaced {
    ... warn
  }
}
```